src/main/java/net/minestom/server/instance/light/starlight/BlockStarLightEngine.java

```diff
@@ -0,0 +1,280 @@
+
+    @Override
+    protected int calculateLightValue(final Instance lightAccess, final int worldX, final int worldY, final int worldZ,
+                                      final int expect) {
+        final Block centerState = this.getBlockState(worldX, worldY, worldZ);
+        int level = centerState.registry().lightEmission() & 0xF;
+
+        if (level >= (15 - 1) || level > expect) {
+            return level;
+        }
+
+        final int sectionOffset = this.chunkSectionIndexOffset;
+//        final Block conditionallyOpaqueState;
+        int opacity = centerState.registry().opacity();
+
+        if (opacity == -1) {
+//            this.recalcCenterPos.set(worldX, worldY, worldZ);
+//            opacity = centerState.getLightBlock(lightAccess.getLevel(), this.recalcCenterPos);
+//            if (((ExtendedAbstractBlockState)centerState).isConditionallyFullOpaque()) {
+//                conditionallyOpaqueState = centerState;
+//            } else {
+//                conditionallyOpaqueState = null;
+//            }
+            return level;
+        } else if (opacity >= 15) {
+            return level;
+        } else {
+//            conditionallyOpaqueState = null;
+        }
+        opacity = Math.max(1, opacity);
+
+        for (final AxisDirection direction : AXIS_DIRECTIONS) {
+            final int offX = worldX + direction.x;
+            final int offY = worldY + direction.y;
+            final int offZ = worldZ + direction.z;
+
+            final int sectionIndex = (offX >> 4) + 5 * (offZ >> 4) + (5 * 5) * (offY >> 4) + sectionOffset;
+
+            final int neighbourLevel = this.getLightLevel(sectionIndex, (offX & 15) | ((offZ & 15) << 4) | ((offY & 15) << 8));
+
+            if ((neighbourLevel - 1) <= level) {
+                // don't need to test transparency, we know it wont affect the result.
+                continue;
+            }
+
+            final Block neighbourState = this.getBlockState(offX, offY, offZ);
+            if (neighbourState.registry().isConditionallyFullOpaque()) {
+//                // here the block can be conditionally opaque (i.e light cannot propagate from it), so we need to test that
+//                // we don't read the blockstate because most of the time this is false, so using the faster
+//                // known transparency lookup results in a net win
+//                this.recalcNeighbourPos.set(offX, offY, offZ);
+//                final VoxelShape neighbourFace = neighbourState.getFaceOcclusionShape(lightAccess.getLevel(), this.recalcNeighbourPos, direction.opposite.nms);
+//                final VoxelShape thisFace = conditionallyOpaqueState == null ? Shapes.empty() : conditionallyOpaqueState.getFaceOcclusionShape(lightAccess.getLevel(), this.recalcCenterPos, direction.nms);
+//                if (Shapes.faceShapeOccludes(thisFace, neighbourFace)) {
+//                    // not allowed to propagate
+//                    continue;
+//                }
```

Comment:
I remember this being the stuff I couldn't port when I tried porting Starlight to Krypton. I wonder if there is a way we can still implement this though, since we can get the shape from extracted data, though there is the question of whether it's even worth it at all. `Shapes.faceShapeOccludes` has some decently complicated logic that would be annoying to implement. If we decide not to implement this functionality at all, there should be comments left here to explain why it's not implemented.